BlueprintUI/Sources/Layout/LayoutWriter.swift

```diff
@@ -0,0 +1,223 @@
+            let result = child()
+            self.add(with: result.0, child: result.1)
+        }
+        
+        /// Enumerates each of the children, allowing you to modify them in place,
+        /// eg to align them all along a common alignment axis or to set a uniform size.
+        public mutating func modifyEach(using change : (inout Child) -> ()) {
+            self.children = children.map {
+                var updated = $0
+                change(&updated)
+                return updated
+            }
+        }
+    }
+    
+    /// Provides the relevant information about the context in which the layout is occurring.
+    public struct Context {
+        
+        /// The size constraint the layout is occurring in.
+        public var size : SizeConstraint
+        
+        /// The environment the layout is occurring in.
+        public var environment : Environment
+    }
+    
+    /// Controls the sizing calculation of the custom layout.
+    public enum Sizing : Equatable {
+        
+        /// Ensures that the final size of element is large enough to fit all children, starting from (0,0).
+        case unionOfChildren
+        
+        /// Fixes the layout size to the provided size.
+        case fixed(CGSize)
+    }
+    
+    /// A child of the custom layout, providing its frame and element.
+    public struct Child {
+        
+        /// The frame of the element in the coordinate space of the custom layout.
+        public var frame : CGRect
+        
+        /// The element to be displayed.
+        public var element : Element
+        
+        /// Creates a new child element.
+        public init(
+            frame : CGRect,
+            element : Element
+        ) {
+            self.frame = frame
+            self.element = element
+        }
+    }
+}
+
+
+extension LayoutWriter {
+    
+    /// We bounce to an inner element so we can provide the environment.
+    private struct InnerElement : Element {
+        var builder : Builder
+        
+        // MARK: Element
+        
+        var content: ElementContent {
+            ElementContent(layout: Layout(builder: self.builder)) { builder in
+                for child in self.builder.children {
+                    builder.add(element: child.element)
+                }
+            }
+        }
+        
+        func backingViewDescription(bounds: CGRect, subtreeExtent: CGRect?) -> ViewDescription? {
+            nil
+        }
+        
+        // MARK: Layout
+        
+        private struct Layout : BlueprintUI.Layout {
+            var builder : Builder
+            
+            func measure(in constraint: SizeConstraint, items: [(traits: (), content: Measurable)]) -> CGSize {
+                switch builder.sizing {
+                case .unionOfChildren:
+                    return builder.children.reduce(CGRect.zero) { rect, child in
+                        rect.union(child.frame)
```

Comment:
I think there's some odd behavior here if any frames have negative values. The element's size will be large enough to bound everything, but the frames won't be placed inside that space when they're laid out.

I think you might want to do one of these, depending on desired behavior:

* In `layout`, offset each frame by the minX/minY of all frames, to normalize the origin. This is probably the expected thing?
* Here, clamp frames to positive values. This would allow negative frames to lay out outside of the element's bounds, but the size is not really the union anymore (e.g. if all children are negative the size would be zero).

Reply:
Hmmm.... interesting. Let me do a couple examples with negative frames. I kinda feel like if you make a frame negative it should just appear starting at that negative frame outside the bounds (since it is an arbitrary layout), and I thought that's what would happen here, but will check and add screenshots.

Reply:
![image](https://user-images.githubusercontent.com/327847/102451903-45c16280-3fee-11eb-9d00-b6661fc1b4ad.png)

So doing negative frames results in this kinda thing; which lines up with other elements like Row and Column that kinda just let their children lay out outside their bounds – which seems fine? You have to try pretty intentionally to end up in this state, and then the layout does what you ask, so I'm not worried about it.

Reply:
> if you make a frame negative it should just appear starting at that negative frame outside the bounds

That seems fine, but what's the measured size in this case? I think you'd want the measured size to exclude content outside the positive area, or else you'll have extra space to the top and right.

Reply:
Updated the docs and added an example here.